sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -452,6 +452,17 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        // A partitioned relation's schema can be different from the input logicalPlan, since
+        // partition columns are all moved after data columns. We Project to adjust the ordering.
+        val input = if (parts.nonEmpty) {
+          val (inputPartCols, inputDataCols) = child.output.partition { attr =>
+            parts.contains(attr.name)
+          }
+          Project(inputDataCols ++ inputPartCols, child)
+        } else {
+          child
+        }
```

Comment:
If we use the name based resolution, we also need to check if the all the input columns have the expected partitioning names; Otherwise, the result will be not predictable.